docs/implementation.adoc

```diff
@@ -0,0 +1,507 @@
+** the information URL of the source
+** the avatar of the source
+3. `jenkins.scm.api.SCMNavigatorEvent` of type `UPDATED` representing the change of metadata for a collection of sources as an aggregate, such as:
+** the description of the collection
+** the display name of the collection
+** the information URL of the collection
+** the avatar of the collection
+
+Implementations are free to use the event system to publish other events as appropriate providing the type of event is logically mapped.
+
+The next step in implementing the SCM API is to allow for consuming plugins to perform deeper identification of interesting `jenkins.scm.api.SCMHead` instances.
+Consuming plugins may not be interested in every single `jenkins.scm.api.SCMHead`.
+For example:
+
+* the https://wiki.jenkins-ci.org/display/JENKINS/Pipeline+Multibranch+Plugin[Pipeline Multibranch Plugin] is only interested in `jenkins.scm.api.SCMHead` instances that have a `Jenkinsfile` in the root of the checkout.
+
+* the https://wiki.jenkins-ci.org/display/JENKINS/Literate+Plugin[Literate Plugin] is only interested in `jenkins.scm.api.SCMHead` instances that have a marker file (configurable with the default being `.cloudbees.md`) in the root of the checkout.
+
+Each SCM API consuming plugin defines the criteria by implementing `jenkins.scm.api.SCMSourceCriteria`.
+Each `jenkins.scm.api.SCMSourceOwner` can specify the criteria for the `jenkins.scm.api.SCMSource` instances that it owns.
+
+When a `jenkins.scm.api.SCMSource` has been supplied with a `jenkins.scm.api.SCMSourceCriteria` it will need to provide a `jenkins.scm.api.SCMProbe` when identifying potential `jenkins.scm.api.SCMHead` instances.
+
+[NOTE]
+====
+Implementations of `jenkins.scm.api.SCMProbe` should put primary focus on the `stat(_path_)` method that checks for existence and filetype.
+This method is expected to be a hot code path
+
+When a consuming plugin is processing a `jenkins.scm.api.SCMHeadEvent`, it will be necessary for the consuming plugin to revalidate the head instance against the `jenkins.scm.api.SCMSourceCriteria`.
+Even if we could trust the content of the `jenkins.scm.api.SCMHeadEvent`, the revalidation would be a necessary step as we have no way of determining from a `jenkins.scm.api.SCMHeadEvent` whether the change has:
+
+* Transitioned the criteria matching from `false` to `true`;
+* Maintained the criteria matching in its current state; or
+* Transitioned the criteria matching from `true` to `false`.
+
+Thus *every* SCM API consuming plugin that listens for a `jenkins.scm.api.SCMEvent` will need to validate that event *against the criteria*.
+This revalidation would be necessary even if we received a signed trusted event from the backing source control system.
+This is why:
+
+* The `jenkins.scm.api.SCMEvent` is always considered `Untrusted` - consuming plugins will be doing revalidation anyway, so it is simpler for consuming plugins to treat everything from an event as untrusted.
```

Comment:
I read this paragraph three times and I'm still fuzzy on the details of what you are trying to say.